lib/dl_file_uploader_worker_lib/dl_file_uploader_worker_lib/tasks/download.py

```diff
@@ -340,3 +347,80 @@ async def run(self) -> TaskResult:
         finally:
             await usm.close()
         return Success()
+
+
+@attr.s
+class DownloadYaDocumentsTask(BaseExecutorTask[task_interface.DownloadYaDocumentsTask, FileUploaderTaskContext]):
+    cls_meta = task_interface.DownloadYaDocumentsTask
+
+    async def run(self) -> TaskResult:
+        dfile: Optional[DataFile] = None
+        redis = self._ctx.redis_service.get_redis()
+
+        try:
+            rmm = RedisModelManager(redis=redis, crypto_keys_config=self._ctx.crypto_keys_config)
+            dfile = await DataFile.get(manager=rmm, obj_id=self.meta.file_id)
+            assert dfile is not None
+
+            assert isinstance(dfile.user_source_properties, YaDocumentsUserSourceProperties)
+
+            yadocs_client = YaDocumentsClient()
+
+            if dfile.user_source_properties.public_link is not None:
+                spreadsheet_ref = yadocs_client.get_spreadsheet_public(link=dfile.user_source_properties.public_link)
+                spreadsheet_meta = yadocs_client.get_spreadsheet_public_meta(
+                    link=dfile.user_source_properties.public_link
+                )
+
+            elif (
+                dfile.user_source_properties.private_path is not None
+                and dfile.user_source_properties.oauth_token is not None
+            ):
+                spreadsheet_ref = yadocs_client.get_spreadsheet_private(
+                    path=dfile.user_source_properties.private_path, token=dfile.user_source_properties.oauth_token
+                )
+                spreadsheet_meta = yadocs_client.get_spreadsheet_private_meta(
+                    path=dfile.user_source_properties.private_path, token=dfile.user_source_properties.oauth_token
+                )
+            else:
+                raise exc.DLFileUploaderBaseError()
+            dfile.filename = spreadsheet_meta["name"]
+
+            s3 = self._ctx.s3_service
+
+            async def _chunk_iter(chunk_size: int = 10 * 1024 * 1024):
+                async with aiohttp.ClientSession() as session:
+                    async with session.get(spreadsheet_ref) as response:
+                        assert response.status == 200
```

Comment:
You don't `assert` behaviour of an external service
Errors must be handled everywhere where we request something from ya-docs API and translated to our exceptions (iirc there are even error descriptions in their responses)

Reply:
> Errors must be handled everywhere where we request something

This hints that it should be a centralized solution (check out aiogoogle error handling – it's a bit clunky because google errors have a complex structure, but it should give you the idea)

https://github.com/datalens-tech/datalens-backend/blob/7482ce0b8dc609d9ccc2d044b6c3b063d9ade63e/lib/dl_file_uploader_lib/dl_file_uploader_lib/gsheets_client.py#L68-L68